Cargo authenticateing users without sending secrets over the network

Eh2406 opened the pull request:
This is spun out of #3139. This provides a way to use asymmetric cryptography to authenticate to registries in a manner flexible enough to be used for *all* communication with a registry.
This could allow a registry to require interaction with a YUBikey to publish. This could allow a registry to use client certificates as a proof of identity. The details of how registries choose to trust their users is out of scope for this RFC, as is figuring out how crates.io could use this system. If you're excited to find ways to use the system, then let's stay focused here and get it merged so you can start experimenting!

[**rendered**](https://github.com/Eh2406/rfcs/blob/master/text/0000-cargo-asymmetric-tokens.md)

kornelski commented:
On macOS, this could be done more securely using its Keychain APIs. 

In macOS there's a chain of trust established between developer of the application that uses credentials all the way to the kernel that checks which codesigned processes can get credentials. Certificates are never written to disk unprotected, and 3rd party applications can never get other apps' credentials. `credential-helper` breaks this chain of trust, and `private-key-path` is much less secure than storage of private keys in the keychain. macOS can protect credentials form code execution with user privileges, this scheme can't.

bjorn3 commented:
> credential-helper breaks this chain of trust

The paseto token returned by the credential-helper can only be used for uploading a single package if I understand the RFC correctly. This makes an attacker running `cargo upload` if cargo were to handle the keychain handling and running the credential helper effectively equivalent in terms of damage the attacker can do. Without this RFC the credential helper could release a long living secret in which case it would be beneficial for cargo to handle the keychain access, but this RFC makes it only release a single use secret.

> If the operation is a mutation, that the package, version, and hash match the request.

Note also that for hardware tokens the "private-key-path" would probably point to some information that needs to be passed to the hardware token and would be useless without the hardware token rather than the actual private key.

tarcieri commented:
One of the benefits of using an expressive bearer token format is being able to set policies around / attenuate those tokens, which is useful in a delegation context like allowing a bot to publish a specific release of a specific crate. See: https://github.com/rust-lang/rfcs/pull/2947

> Another point on macOS (which would probably be addressed by what @kornelski suggests) is that it would be great to leverage the TouchID support that many modern macOS machines have.

TouchID is ultimately an item on an AuthZ ACL you can configure for a particular keychain item as stored by Keychain Services (possibly in the SEP, if that's desired). FWIW, I wrote a crate with bindings to the relevant API surface which would be nice to upstream into the `security-framework` crate: https://keychain-services.rs/

The tricky part @kornelski mentioned is this kind of feature would need to be directly added to the cargo binary itself, as there is currently no way to e.g. use X.509 codesigning certificate chains to delegate authority to use TouchID to a signed `credential-helper` (I have asked Apple security engineers about this and they confirmed this isn't possible).

Another problem is if you do want to store keys in the Secure Enclave Processor (SEP), they must be 256-bit. For ECDSA this means using the NIST P-256 elliptic curve, but PASETO only supports NIST P-384.

est31 commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,183 @@
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In [`config.toml`](https://doc.rust-lang.org/cargo/reference/config.html) and `credentials.toml` files there is a field called `private-key-path`, which provides a path to a `PKCS#12` formatted file containing a private key.
```
I think storing the key directly in credentials.toml makes most sense, in base64 format or similar. Most users won't benefit from having it in a separate file. Instead, there is the burden of having *two* files that need to be protected, and it also means more complexity. Custom use cases can use a credentials process which is superior I think.

tarcieri replied:
I would suggest using PKCS#8 instead of PKCS#12. The latter is a container format for both PKCS#8 private keys (which are themselves asymmetric keypairs) and X.509 certificates, and X.509 certificates don't seem relevant to this particular application.

One nice thing about using PKCS#8 as opposed to something like credentials.toml is support for password-protected encrypted private keys, which can be used to protect these keys at rest. PKCS#8v2 supports deriving an encryption key using PBKDF2 or scrypt and using that to encrypt the original private key using AES.

est31 replied:
You can use `PKCS#8` or whatever other format, but ideally you don't store things into *yet another* file. That annoys users. toml supports multiline strings, so you can even have PEM encoding if you want.

Eh2406 replied:
I was already convinced that cargo should only support one way of getting the key, and that all other ways of storing keys should require a credential process. I am now convinced that cargo should accept the (encoded) key directly as opposed to a path to the key.

The argument that convinced me is that most CI/CD systems allow you to store secrets in environment variables, but not particularly in files. If cargo accepts the key directly, then we already have an environment variable that will go straight from CIs secrets system to cargo. The same way it does for the existing secret tokens.

I did not have time to edit this into the RFC today, but I will do so and my next opportunity.

Diggsey commented:
> The file containing the token can be read at rest. File permissions are used to protect it, but can only go so far. [Credential processes](https://github.com/rust-lang/rfcs/blob/161ce8a26e70226a88e0d4d43c7914a714050330/text/2730-cargo-token-from-process.md) can do better if they are used.

Of the motivations for this feature, this seems by far the strongest to me. Package managers like cargo are particularly vulnerable to this kind of attack because malicious build scripts can easily exfiltrate such files. The other possible attacks seem much more theoretical, or else can be protected against without user-facing changes.

For that reason, I don't think that an overall push to use private keys (stored at rest) instead of tokens (stored at rest) is justified. For example, if I'm setting up a publish CI pipeline, a token is more convenient than a private key, yet a private key does not protect against the primary risk (CI secret being leaked), and given how frequent setting up a CI pipeline is for me, any additional complexity for no benefit is just going to waste a lot of my time.

> If a registry does not adequately protect its copy of the tokens then a database disclosure can leak all the users' tokens.

Even with tokens, there's no reason for a registry to store tokens in a recoverable format. The registry can store token hashes instead, and avoid this problem without inconveniencing the user. Since tokens are already random, you can avoid a lot of the complexities of storing passwords.

However, specifically for accounts used by actual humans, then I would support a push to any solution that avoids storing keys at rest (ie. using an agent or hardware token) because in that case there is a much greater potential benefit.

est31 commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,183 @@
+- print the public key and the path to the file.
+  (See unresolved questions section.)
+
+There is also a field called `user-id` which is a string chosen by the registry, which is intended to be non-secret and used for identifying the user. Cargo requires it to be non-whitespace printable ASCII, registries should base64 encode non-ASCII data. Cargo may broaden the allowed set of characters to some other subset of UTF-8 in the future.
```
I'm not sure a `user-id` is a good idea. Especially as people change can change usernames and sometimes want their old names gone from the records, so ideally one designs technology that does without usernames. I guess the purpose of the `user-id` is to figure out which public key the private key corresponds to? For this purpose, sending a *key* identifier would make more sense, as cargo supports multiple tokens per user. The key identifier would be purely derived from the key so the user wouldn't even have to enter it, which reduces setup overhead.

Eh2406 replied:
I really like the idea of using a "key ID", say the SHA-384 of the public key! This removes a step for most use cases. For the rare use cases where the server does want some arbitrary information to be passed back as part of the token we can add a field for that purpose. (Here is cryptographic proof that the central CA server authorized this action. Or whatever this particular registry needs.) That field can have a less confusing name. I did not have time to edit this into the RFC today, but I will do so and my next opportunity.

tarcieri replied:
A standard way of encoding that information is via [SPKI key fingerprints](https://docs.rs/spki/latest/spki/struct.SubjectPublicKeyInfo.html)

est31 commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,183 @@
+In practice, I suspect many registries will not, leading to an ecosystem where most registries use less secure authentication, and creating more hazards for users. Some security properties (e.g. not supporting tokens from one registry on another) work better when all registries support them.
+
+We could use PASETO `v4.public` instead of `v3.public`. The `v4` standard uses more modern crypto, based on Ed25519.
+Unfortunately, most existing hardware tokens do not support Ed25519. By using `v3.public` based on P-384 we allow a `credential-process` to keep keys on the hardware.
```
I found some comparison on hn, which says that ed25519 is generally better: https://news.ycombinator.com/item?id=19749452

Furthermore, ed25519 has less red stuff at the table at [safecurves](http://safecurves.cr.yp.to/index.html) compared to the NIST curves.

Generally it might be a good idea to support multiple schemes from the start, to ensure that there is algorithm agility.

tarcieri replied:
> Furthermore, ed25519 has less red stuff at the table at [safecurves](http://safecurves.cr.yp.to/index.html) compared to the NIST curves.

The table on the safecurves site is critiquing the obsolete and incomplete formulas I was describing earlier, and is specifically addressed in the paper I linked (p.5), where citation `[14]` is the safecurves site:

> Prime order curves can be safe. Several of the standardized prime order curves mentioned
above have recently been critiqued in [14], where they were deemed not to meet (some or all
of) the four “ECC security” requirements: (i) Ladder, (ii) Twists, (iii) Completeness, and (iv)
Indistinguishability.
> On the contrary, this paper shows that prime order curves have complete formulas that are
comparably efficient.

Why the safecurves site has not been updated to reflect the developments of this paper is uncertain, however it contains obsolete information.

The security properties of implementations of Edwards curves as used by Ed25519 or short Weierstrass curves like NIST P-256 and P-384 implemented using the complete projective formulas are effectively the same, and algorithmically quite similar. As it were, the RustCrypto [`p256`] (https://github.com/RustCrypto/elliptic-curves/tree/master/p256) crate is implemented using these formulas, and we are working implementing [`p384`](https://github.com/RustCrypto/elliptic-curves/tree/master/p384) with them as well.

Where Ed25519 benefits is areas like performance (Edwards curves are more amenable to parallel implementations) and more exotic applications like batch verification, multisignatures, and signature aggregation, which are outside of the scope of this particular use case.

> Generally it might be a good idea to support multiple schemes from the start, to ensure that there is algorithm agility.

I would suggest supporting Ed25519 along with ECDSA/P-{256,384} (possibly only choosing one of P-256 or P-384). While Ed25519 is great, support for it in things like hardware tokens, OS keychains, HSMs, and cloud KMS services is still somewhat spotty, but slowly improving.

RalfJung replied:
To make sure I understand correctly, the RFC as-is currently would *not* support Ed25519, and supporting both v3 and v4 (and thus both P-384 and Ed25519) is not part of the RFC? But the RFC is written in a way that that would be a possible extension in the future?

Eh2406 replied:
I am not a cryptographer and am new to all the details here. So I'm going to explain as carefully as I can, so that someone can tell me if I have something wrong.

`PASETO V4` is an encoding that requires the use of the algorithm `Ed25519`.
`PASETO V3` is an encoding that requires the use of the algorithm `ECDSA/P-384`.
Also all the PASETO variations clearly identify what version they came from.

The RFC as currently written only supports `PASETO V3`. That means that cargo will only understand how to generate and read keys for `PASETO V3` and use the algorithm `ECDSA/P-384`. Furthermore it will insist that credential processes return something that appears to be correctly formatted as a `PASETO V3`.

It is not hard to imagine how this could be lifted in a future RFC. This would allow us to allow for a new format, say `PASETO V4` or `Biscuit`.
That RFC would have to specify:
- How `--generate-keypair` decides `V3`/`V4`/`Biscuit`? Presumably, a new flag to specify format and a timeline for the default to change.
- When reading a config file how cargo decides which one it has found? Almost certainly, read the version number encoded in the key and use its matching format.
- When receiving from a credential process which style to check? Presumably, we allow anything that looks like any of the supported formats. (Note that this could be done before and independently of the other changes.)
- Whether to deprecate and remove the old format? This would require extreme justification, but maybe worth discussion depending on the urgency of switching.
- Whether and how a registry can request a particular format? Presumably, Cargo will send whatever format is specified in the config, if it is not a format supported by this registry the registry will return an error code to show the problem. (How does this work with credential providers?)

Does that answer your questions? Did I miss something important?

RalfJung replied:
Yes, that all makes sense, thanks. I am not clear on how 'format detection' would be done (are all these formats clearly distinct from each other in their encoded form, do we use filenames, whatever), but as long as that option exists I do not care about the technical details. :)

Eh2406 replied:
> are all these formats clearly distinct from each other in their encoded form?

Basically yes. Tokens encoded in `PASETO V3` start with `v3.public.`. `PASETO V3` keys encoded in `PASERK` start with `k3.secret.` and key IDs start with `k3.pid.`. (That sentence is also true if you replace all the `3`s with `4`s.) If we are switching to a different family of formats, which do not have a clear prefix to identify them, we can define that we are sticking some arbitrary string in front for disambiguation purposes.

Eh2406 commented:
I added some language addressing UX concerns.

When https://github.com/rust-lang/rust/issues/27694 has been completed I would strongly approve of an RFC moving to using the keychain for storage on MacOS. Given that https://github.com/rust-lang/rust/issues/27694 is not going to be closed imminently I think discussion of what we could do after it is out of scope for this RFC. Especially with the explicit section "Note on stability" warning people that if more secure things come along this RFC may be superseded.

I added text discussing [Amazon's SigV4](https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html). Thank you for reminding me to discuss it.

I'm not clear why JOSE is simpler then PASETO, they are theoretically very similar formats for connecting some data to signatures of that data. JOSE can describe many more algorithms and methods, requiring documents like the one you linked to spend their time describing what parts they are comfortable using. Each version of the PASETO is only compatible with one algorithm. I don't know anything about ACME. @djc, Can you explain what aspect of what they're doing you find simpler?

> For example, if I'm setting up a publish CI pipeline, a token is more convenient than a private key, yet a private key does not protect against the primary risk (CI secret being leaked), and given how frequent setting up a CI pipeline is for me, any additional complexity for no benefit is just going to waste a lot of my time.

We do need to be cognizant of the user experience! For use cases where a token would be just as secure, we need to keep overhead down. Generating a private key locally and putting it in CIs list of secrets, can hopefully be only one small additional step then putting the token in the secrets.

> Even with tokens, there's no reason for a registry to store tokens in a recoverable format. 

You're absolutely correct. This is only a problem in practice not in theory. However, the sentence you are responding to links to an example of crates.io getting this wrong. I can only assume if we have seen one registry get this wrong, then there are others and there will be more in the future.

> However, specifically for accounts used by actual humans, then I would support a push to any solution that avoids storing keys at rest (ie. using an agent or hardware token) because in that case there is a much greater potential benefit.

We have to make it possible if we wanted to become widely adopted. I think this massively increases the ability for storing the critical secrets on dedicated hardware.

djc commented:
> @djc, Can you explain what aspect of what they're doing you find simpler?

It's mostly the notion of tokens vs signing that I was objecting to. I think "token" is loaded with a different semantic meaning than signatures. But this now makes more sense given this comment from @tarcieri:

> One of the benefits of using an expressive bearer token format is being able to set policies around / attenuate those tokens, which is useful in a delegation context like allowing a bot to publish a specific release of a specific crate.

pierwill commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,196 @@
+Different registries will have different users in mind and have different use cases. Therefore, they will need to have different behaviors. So, there are many decisions a registry has to make that this RFC has no opinion on. Some examples:
+
+- Bootstrapping trust: how does the registry decide to trust a new user?
+- Key generation: where and how the key pair is made?
```
```suggestion
- Key generation: where and how is the key pair made?
```

pierwill commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,196 @@
+
+There are credential processes for using key pairs stored on hardware tokens. Check crates.io to see if there's one available for your hardware. Each one is a little different, but the general workflow is:
+1. `cargo install credential-process-for-your-hardware-token`
+2. run `cargo credential-process-for-your-hardware-token setup registryURL` to get your public key.
```
```suggestion
2. Run `cargo credential-process-for-your-hardware-token setup registryURL` to get your public key.
```

pierwill commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,196 @@
+- print the public key and the path to the file.
+  (See unresolved questions section.)
+
+There is also a field called `user-id` which is a string chosen by the registry, which is intended to be non-secret and used for identifying the user. Cargo requires it to be non-whitespace printable ASCII, registries that need non-ASCII data should base64 encode it.
```
```suggestion
There is also a field called `user-id` which is a string chosen by the registry, which is intended to be non-secret and used for identifying the user. Cargo requires it to be non-whitespace printable ASCII. Registries that need non-ASCII data should base64 encode it.
```

pierwill commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,196 @@
+
+- The PASETO is in v3.public format.
+- The PASETO validates using a public key for that user ID.
+- The URL matches the registry base URL (to make sure a PASETO sent to one registry can't be used to authenticate to another, and to prevent typosquatting/homoglyph attacks)
```
```suggestion
- The URL matches the registry base URL (to make sure a PASETO sent to one registry can't be used to authenticate to another, and to prevent typo-squatting/homoglyph attacks)
```

pierwill commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,196 @@
+- The PASETO is in v3.public format.
+- The PASETO validates using a public key for that user ID.
+- The URL matches the registry base URL (to make sure a PASETO sent to one registry can't be used to authenticate to another, and to prevent typosquatting/homoglyph attacks)
+- The PASETO is still within its valid time period (to limit replay attacks) 15 minutes will be recommended. but a shorter time can be used by a registry to further decrease replayability. Or a longer one can be used to better accommodate clock skew.
```
```suggestion
- The PASETO is still within its valid time period (to limit replay attacks). We recommend a 15 minute limit, but a shorter time can be used by a registry to further decrease replay-ability. Or a longer one can be used to better accommodate clock skew.
```

pierwill commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,196 @@
+
+## Credential Processes
+
+Credential Processes as defined in [RFC 2730](https://github.com/rust-lang/rfcs/pull/2730) are outside programs cargo can call on to change where and how secrets are stored. That RFC defines `special strings` which go in the `credential-process` field to describe what data the process needs from cargo. This RFC adds `{claims}`. If used Cargo will replace it with a json encoded set of key value pa ...
```
```suggestion
Credential Processes as defined in [RFC 2730](https://github.com/rust-lang/rfcs/pull/2730) are outside programs cargo can call on to change where and how secrets are stored. That RFC defines `special strings` which go in the `credential-process` field to describe what data the process needs from cargo. This RFC adds `{claims}`. If used Cargo will replace it with a JSON encoded set of key value pairs that should be in the generated token. Cargo will check that the output of such a process looks like a valid PASETO v3.public token that Cargo would have generated, and that the PASETO token includes all the claims Cargo provided. The credential process may add additional claims (e.g. 2fa, TOTP), as long as they are nested in `custom`.
```

Eh2406 commented:
I did a major round of editing on Thursday but forgot to write a summary.
The changes were:
- Clarify how I'm using the terms "secret tokens" vs. "asymmetric tokens" vs. "hardware token".
- Use a standard "key ID" instead of a "user ID" removing a step for most users and a confusing piece of terminology.
- Add a "key-subject" for cases where additional data is needed to identify the key aside from the "key ID".
- Describe what would be needed for algorithm agility as a future possibility.
- Acknowledge that "secret tokens" do not *need* to be stored in plain text on the server.
- Move the private token to `credentials.toml` in `PASERK` for better compatibility with CI.

arlosi commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,218 @@
+
+Some registries prioritize user experience over strictest security. They can simplify the process by providing key generation on the server. If your registry works this way the workflow will be:
+1. Log into the registries website
+2. Go to the "register generate a key pair" page, and copy the command it generated for you. It will disappear when you leave the page, the server will not keep a copy of the public key!
```
Should this be "the server will not keep a copy of the **private** key"?

Eh2406 replied:
Thank you for the correction.

arlosi commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,218 @@
+Maven Central [requires](https://maven.apache.org/repository/guide-central-repository-upload.html#pgp-signature) all uploads to be [signed by PGP](https://central.sonatype.org/publish/requirements/gpg/) and that the keys are registered with a public key server. Following the UNIX philosophy, they leave the actual act of signing up to independent implementations of PGP.
+It takes a lot of documentation to explain how to hook up all of these different parts to work together correctly. Furthermore, no assurance is made that the GPG signature and the Maven Central token used for upload represent the same identity.
+
+The npm client can pass along a `otp` option on the command line to act as [proof of 2FA](https://docs.npmjs.com/configuring-two-factor-authentication#sending-a-one-time-password-from-the-command-line). This provides a lot of the "over the wire" benefits of this RFC for the npm registry, but cannot be used by a third party after the fact to verify the uploaded identity nor is it easy for alternat ...
```
> nor is it easy for alternative registries to set up

TOTP seems simple enough for alternative registries to implement. It's pretty well standardized how to validate TOTP codes. I still prefer the option described in this RFC however.

Eh2406 replied:
Made the mistake of thinking that because I couldn't figure out how to do it it was hard. I removed that part of the sentence.

arlosi commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,218 @@
+
+Both fields can be set with `cargo login --registry=name --private-key="key" --private-key-subject="subject"`.
+
+A registry can have at most one of `private-key`, `token`, or `credential-process` set.
```
Should the private key have some kind of prefix so we can identify it as PASETO v3 in case we ever want to change to a different key format?

Eh2406 replied:
v3 Private keys encoded in PASERK secret format start with `k3.secret.`. So I think it will be easy to tell that the `private-key` was intended for this RFC.

arlosi commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,218 @@
+
+[TUF](https://theupdateframework.io/) exclusively deals with how a client downloading packages through a mirror can be assured they came from a non-compromised copy of the registry. Which is not the problem this RFC is addressing.
+
+# Unresolved questions
```
How do we handle tokens passed on the command line via `--token`. Will we also support passing a private key via command line?

Eh2406 replied:
It is possible to set it using the environment variable. But we have a shorthand for setting `--token` so maybe we should have a shorthand for `private-key` and `private-key-subject`. Do you think it's reasonable to consider this a part of the unresolved question about CLI design, or should we resolve it in the text of this RFC?

arlosi replied:
I think we should make a decision about it in the RFC.

I'd rather avoid passing private keys on the command line since it increases the chances of them being exposed.

Eh2406 replied:
The RFC no longer mentions passing keys on the command line.
The existing commands that take a `--token`, if it is set it will override any `private-key`/`private-key-subject` from the config.toml or environment variables. That's the only thing that maintains backwards compatibility.

bjorn3 commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,218 @@
+
+Some registries prioritize user experience over strictest security. They can simplify the process by providing key generation on the server. If your registry works this way the workflow will be:
+1. Log into the registries website
+2. Go to the "register generate a key pair" page, and copy the command it generated for you. It will disappear when you leave the page, the server will not keep a copy of the public key!
```
The web crypto api allows you to generate the keypair in the browser I believe, which would be a bit safer.

Eh2406 replied:
Generating on the client would be better design. But I am not seeing a way to edit the sentence to make that clearer.

arlosi replied:
You could replace "on the server" with "on the registry website". Then it's up to the web application to decide how to generate keys.

I think it's clear that this is a less-secure option than locally generating the keys.

Eh2406 replied:
`on the server` -> `in the browser` Good call!

bjorn3 commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,218 @@
+Some registries prioritize user experience over strictest security. They can simplify the process by providing key generation on the server. If your registry works this way the workflow will be:
+1. Log into the registries website
+2. Go to the "register generate a key pair" page, and copy the command it generated for you. It will disappear when you leave the page, the server will not keep a copy of the public key!
+3. Run it on the command line. It will look like  `cargo login --registry=name --private-key="key"`
```
This will leak the private key into the shell history. Maybe prompt for the private key instead?

Eh2406 replied:
This is an existing problem with the login command. On the other hand this RFC might be good opportunity to fix it. Do you think it's reasonable to consider this a part of the unresolved question about CLI design?

bjorn3 replied:
Sure

arlosi replied:
I believe if you do "cargo login" without specifying the token, cargo will prompt for it. However, if there are now two ways (token and private key) cargo would need to identify the pasted data as one or the other appropriately.

Eh2406 replied:
I changed this to no longer have `--private-key="key"` but just a `--private-key` that signifies that the prompt will be asking you for a key and not a token.

arlosi commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -0,0 +1,218 @@
+
+Figuring out how and when crates.io should support these kinds of tokens is left to a follow-up discussion/RFC. The motivation section describes some of the things that will need to be figured out.
+
+Only after crates.io is not using secret tokens should we remove the support for them in private registries (and the code in cargo to support it).
```
I really like the idea of using public key crypto to improve cargo security, however I'm somewhat concerned about deprecating any other type of authentication. It may make it harder to integrate cargo into existing authentication systems.

Using a `credential-process` to issue a short-lived token seems like a flexible and secure option. It could work similar to how GitHub does HTTP authentication after installing the [credential provider](https://github.com/GitCredentialManager/git-credential-manager). Cargo would call the credential provider with the existing information (registry url, etc), the credential provider could verify identity any means it wanted (OAuth popup, 2FA key entry, etc), then produce a short-lived bearer token for cargo to send.

Eh2406 replied:
I think you're reminding us that it is possible to use secret tokens with `credential-process` in a secure way. Which is definitely an important thing to keep in mind.

With some ingenuity, you can use a credential-process to smuggle a secret token into the format of a asymmetric token. The credential process creates a brand-new key pair, puts in "custom" the secret token and the public key. The server can validate that the asymmetric token is valid according to this RFC, but then makes all AuthN decisions based on the embedded secret token. The decision about removal is about how much work is involved and for whom, not what is possible.

Fundamentally, this is in the future possibilities section. This RFC does not suggest deprecation or removal. The time to discuss whether deprecation or removal is a good idea is in future RFC that suggests making that change.

bjorn3 commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -63,7 +63,7 @@ There are credential processes for using key pairs stored on hardware tokens. Ch
 1. Log into the registries website
 2. Go to the "generate a key pair" page, and copy the command it generated for you. It will disappear when you leave the page, the server will not have a copy of the private key!
-3. Run it on the command line. It will look like  `cargo login --registry=name --private-key="key"`
+3. Run it on the command line. It will look like  `cargo login --registry=name --private-key` witch will prompt you to put in the key value.
```
*which

bjorn3 commented on text/0000-cargo-asymmetric-tokens.md:
```diff
@@ -83,7 +83,7 @@ There is also an optional field called `private-key-subject` which is a string c
 It is intended for the rare use cases like "cryptographic proof that the central CA server authorized this action". Cargo requires it to be non-whitespace printable ASCII. Registries that need non-ASCII data should base64 encode it.
 
-Both fields can be set with `cargo login --registry=name --private-key="key" --private-key-subject="subject"`.
+Both fields can be set with `cargo login --registry=name --private-key --private-key-subject="subject"` witch will prompt you to put in the key value.
```
*which

ehuss commented:
I'm going to move this forward to the final comment period.  I think this represents a good step forward for registry authentication. Thanks @Eh2406 for putting all the work into the RFC!

@rfcbot fcp merge

tarcieri commented:
> If I've misunderstood the situation, please feel free to show how current Biscuits already can implement all of those properties, not just that it may be able to in the future.

Can you highlight which specific concerns you have with Biscuits?

Eh2406 commented:
Biscuits as they are now only have one secret key. The secret key used to create the root biscuit. From there on out if you have the biscuit you have all that is required to use it. (Yes, internally secret keys are involved, but there always included in the message. So the point stands.) 

The recommended way to use biscuits is for this secret key to be kept server-side. The server attenuates down this token before sending it to the user. Anyone who has this token can use it for anything it is attenuated to do. This directly falls afoul of the first two "Key additional design properties" I linked to.

The server could add an attenuation that requires some other form of AuthN in order to use the token before sending it to the user. At which point this RFC can be that "other form of AuthN".

Geal commented:
from what I understand, you're going towards a design where the client holds a secret key and uses it to sign something so that no secret info goes over the wire and is shared with the server?

This is compatible with Biscuit's design:
- the client generates a root key pair
- on registration, the client gives the public key to the server
- the client creates a token. That token can contain info like the registry URL
- request are sent with the token and a user id (used to look up the public key), the token is verified, data is loaded, and the authorizer has checks for the URL, current time, etc
- now there's the interesting property: the client can keep the private key in a safe place and use a basic, long lived token locally that is attenuated on every request, possibly with a short expiration time or binding it to a specific operation. And the client can create derived token for any machine, CI, etc that could need access to this registry

joshtriplett commented:
One of the goals of this RFC is to pick a *specific* authentication scheme and token format. We set the requirement early on that we wanted to pick one format and use versioning, and future format changes would occur via new versions. This seems to be the direction that cryptographic work is taking in general: selecting specific algorithms and formats rather than being arbitrarily extensible.

As a result, our goal is to pick a specific format that meets all the requirements. Our goal is also to pick a format that we can get a security expert to take a look at and confirm that it meets those requirements. And all else being equal, we're looking for the simplest solution that meets those requirements. We've currently done all of those things for the current selection of PASETO.

If another format exists that may potentially meet the requirements, that alone isn't enough to argue that we should switch to that format.

To consider switching to a different format at this point, we would need 1) to be confident that it met all the requirements, in detail, 2) to have security expertise available that can review that format in detail, and 3) to have a specific valuable reason to switch above and beyond both of those (since 1 and 2 would just confirm that another format is viable, not that another format is better). And the explanation for (3) would need to show that that feature is something we actively want *now* (not just in the hypothetical future) and that it's worth the complexity added.

I should also mention that this choice is not irrevocable; if there's an issue in the future, or if there's a feature we need and need to switch formats to get, we can always switch using a new version number.

ehuss commented:
**Huzzah!** The @rust-lang/cargo team has decided **to accept** this RFC.

To track further discussion, subscribe to the tracking issue here:
https://github.com/rust-lang/cargo/issues/10519

bjorn3 commented on text/3231-cargo-asymmetric-tokens.md:
```diff
@@ -1,7 +1,7 @@
-- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
-- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+- RFC PR: [rust-lang/rfcs#3231](https://github.com/rust-lang/rfcs/pull/3231)
+- Cargo Issue: [rust-lang/rust#10519](https://github.com/rust-lang/cargo/issues/10519)
```
```suggestion
- Cargo Issue: [rust-lang/cargo#10519](https://github.com/rust-lang/cargo/issues/10519)
```

slanterns commented:
Also, the rendered link is broken.

ehuss commented:
Thanks, I have fixed both issues.
